File: python/matrix/matrix.py
''' Matrix handler class.

Given a string representing a matrix of numbers,
return the rows and columns of that matrix.
'''


class Matrix:
    '''Matrix class, returns row and columns of matrix.'''

    def __init__(self, matrix_string: str):
        '''Initializer of Matrix class.'''
        self.matrix = []
        for row in matrix_string.split("\n"):
            self.matrix.append([int(num) for num in row.split()])

    def row(self, index: int) -> list[int]:
        '''Returns row given by index.'''
        # It is safer to return a copy of the row, in case you change it.
        return self.matrix[index-1].copy()

    def column(self, index: int) -> list[int]:
        '''Returns column given by index.'''
        col = []
        for row in self.matrix:
            col.append(row[index-1])
        return col

File: javascript/matrix/matrix.js
/**
 * Given a string representing a matrix of numbers, return the rows and columns
 * of that matrix.
 * @file
 * @author Andreas Atle, atle.andreas@gmail.com
 */

/**
 * Class representation of matrix.
 * @class 
 */
export class Matrix {
  matrix;
  transpose;
  
  /** 
   * @constructor
   * @param {string} matrix String representation of a matrix.
   */
  constructor(matrix) {
    this.matrix = [];
    // Alternative way of writing a for-loop in js.
    matrix.split('\n').forEach((line, i) => {
      let row = [];
      line.split(' ').forEach((val, j) => {
        row.push(Number(val));
      })
      this.matrix.push(row);
    })

    // Compute the transpose of the matrix.
    this.transpose = [];

    for (let j = 0; j < this.matrix[0].length; j++) {
      this.transpose[j] = [];
    }

    for (let row of this.matrix) {
      for (let j = 0; j < row.length; j++) {
        this.transpose[j].push(row[j]);
      }
    }
  }

  /**
   * Get the rows of the matrix.
   * @public
   * @method
   * @returns {number[][]}
   */
  get rows() {
    return Object.assign({}, this.matrix);
  }

  /**
   * Get the columns of the matrix.
   * @public
   * @method
   * @returns {number[][]}
   */
   get columns() {
    return Object.assign({}, this.transpose)
  }
}
